### ISSUE TITLE
[C++][Parquet] Add ability to write/read repetition/definition levels with PLAIN encoding

### ISSUE THREAD

**Issue by Hattonuri** [apache/arrow#39581]
### Describe the enhancement requested

As for today, the biggest parts of reading parquets in my code look like that
<img width="1723" alt="image" src="https://github.com/apache/arrow/assets/53221537/afb26b31-dac1-47ba-b7e1-215d7b6b135c">
It is clearly seen that reading values takes less time than reading levels
Data is written using zstd and I would assume that decompression of levels using ready compressor would be much better in speed while preserving the same compression ratio

### Component(s)

C++, Parquet

**Comment by mapleFU** [comment]
def/rep level encoding is part of parquet standard, which is not a part of arrow 🤔 Maybe we can trying to making it faster...

**Comment by mapleFU** [comment]
@Hattonuri Maybe this issue is also related: https://github.com/apache/arrow/issues/39227 . We can trying to optimizing decoding the bits here 🤔

**Comment by Hattonuri** [comment]
I think there is hardly any chance to be as fast as ready decompressor :(

**Comment by Hattonuri** [comment]
Probably there will be any magic avx instruction that will unpack bit mark to bytes 👀

**Comment by mapleFU** [comment]
Here we need to impl `unpack16`, I think xsimd or other let compiler help may works.

**Comment by mapleFU** [comment]
```c++
    // TODO: revisit this limit if necessary
    DCHECK_LE(num_bits, 32);
    const int buffer_size = 1024;
    uint32_t unpack_buffer[buffer_size];
    while (i < batch_size) {
      int unpack_size = std::min(buffer_size, batch_size - i);
      int num_unpacked =
          internal::unpack32(reinterpret_cast<const uint32_t*>(buffer + byte_offset),
                             unpack_buffer, unpack_size, num_bits);
      if (num_unpacked == 0) {
        break;
      }
      for (int k = 0; k < num_unpacked; ++k) {
#ifdef _MSC_VER
#pragma warning(push)
#pragma warning(disable : 4800)
#endif
        v[i + k] = static_cast<T>(unpack_buffer[k]);
#ifdef _MSC_VER
#pragma warning(pop)
#endif
      }
      i += num_unpacked;
      byte_offset += num_unpacked * num_bits / 8;
    }
```

Current code use SIMD unpack to 32 bit, and copying to int16_t 🤔I'm not sure can we depend on auto-vectorize

**Comment by mapleFU** [comment]
Also, after re-checking the `RleDecoder`, I think unpack16 might improve the performance but it's not the bottleneck in your case, here maybe the `GetBatch<int16_t>` loop be slow...

**Comment by Hattonuri** [comment]
Probably bit pack can use https://github.com/chenkovsky/TurboPFor
And then write levels with bit pack

**Comment by Hattonuri** [comment]
By the way he also has fast rle encoding https://github.com/powturbo/Turbo-Run-Length-Encoding
But as i understood there are many rles.....

**Comment by emkornfield** [comment]
> def/rep level encoding is part of parquet standard, which is not a part of arrow 🤔 Maybe we can trying to making it faster...

Agree, while its not explicitly stated, even [though the thrift definition allows for specifying encoding](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L585) I don't think plain encoding was ever intended here (would have to take a closer look at parquet-mr).  And since there isn't a native primitive type for [int16](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L32) we'd have to use int32 as the primitive type, which I guess would compress well but still require a cast.

> Probably bit pack can use https://github.com/powturbo/TurboPFor-Integer-Compression
And then write levels with bit pack

I think [original code for bin packing](https://github.com/apache/arrow/blob/e6323646558ee01234ce58af273c5a834745f298/cpp/src/arrow/util/bpacking_default.h#L20)  probably came from that source.

> Also, after re-checking the RleDecoder, I think unpack16 might improve the performance but it's not the bottleneck in your case, here maybe the GetBatch<int16_t> loop be slow...

I think @mapleFU suggestion of having a specialized unpack16 makes sense, I think at the very least this avoids an extra round trip from to/from the cache if we can write directly to the output value instead of having a temporary value holder?

**Comment by mapleFU** [comment]
Let me first draft a benchmark for `RleDecoder` and do later optimization based on that.

Also, would you mind provide levels distribution in your case @Hattonuri

**Comment by emkornfield** [comment]
Also, @Hattonuri the code linked is licensed under GPL, this cannot be used in Apache projects.

**Comment by Hattonuri** [comment]
> Also, would you mind provide levels distribution and read batch size in your case

read batch size is 2048
I am reading not-null arrays(but all parts are optional) of length 20. So definition levels are all 3 and repetition levels are [0] + [1] * 19 repeatedly

**Comment by mapleFU** [comment]
@Hattonuri I've written a benchmark here: https://github.com/apache/arrow/pull/39705

Conclusion: maybe you can try bitpack, and if using bitpack and implement unpack16, it would be fast 🤔

**Comment by mapleFU** [comment]
Yeah, maybe it's ok to add one and testing it's able to read that 🤔

**Comment by thisisnic** [comment]
This issue has been closed because there has been no activity for the 14 days since the stale warning. Feel free to reopen if this has been closed in error.
